cmd/metrics-prometheus-collector/cmd/cmd.go

```diff
@@ -299,18 +319,89 @@ func (ms *MetricsServer) Start(ctx context.Context) error {
 	// this shouldn't be necessary since we shutdown when we aren't the leader, but
 	// serves as a sanity check in case we don't shutdown gracefully and quickly
 
+	go ms.continouslyCacheResponse() // continuously read metrics and cache the result
+	go ms.serveMetricsFromCache()    // serve "/metrics" http requests from the cache
+
 	<-ctx.Done()
 	log.Info("ending sampler server", "reason", ctx.Err())
 	return nil
 }
 
-func (ms *MetricsServer) stopLeading(current_id string) {
-	if exitAfterLeaderElectionLoss {
-		log.Info("exiting after leader election loss")
-		// TODO: we shouldn't actually need to do this, but are seeing issues
-		// where pods never re-aquire leadership after loss.  Kicking the process
-		// to make it try to re-establish leadership.
-		os.Exit(0)
+// cache the response from making a local http request
+type cachedResponse struct {
+	data    []byte
+	headers http.Header
+}
+
+// continouslyCacheResponse continuously hits the collector metrics endpoint and
+func (ms *MetricsServer) continouslyCacheResponse() {
+	t := time.NewTicker(ms.PreComputeMetrics.Frequency)
+	defer t.Stop()
+	for {
+		start := time.Now()
+		backoff := retry.DefaultBackoff
+		backoff.Cap = time.Minute * 5
+		backoff.Duration = time.Millisecond * 100
+		err := retry.OnError(backoff,
+			func(err error) bool { return true },
+			func() error {
+				resp, err := http.Get("http://" + options.MetricsBindAddress + "/metrics")
+				if err != nil {
+					log.Error(err, "unable to read cached metrics response")
+					return err
+				}
+				body, err := io.ReadAll(resp.Body)
+				if err != nil {
+					log.Error(err, "unable to read cached metrics body")
+					return err
+				}
+				resp.Body.Close()
+				ms.cachedResponseBody.Store(cachedResponse{
+					data:    body,
+					headers: resp.Header,
+				})
+				return nil
+			})
+		if err != nil {
+			log.Error(err, "unable to cache metrics by reading local /metrics")
+			ms.stop()
+			os.Exit(1)
+		}
+		log.Info("finished caching metrics response",
+			"seconds", time.Since(start).Seconds())
+		collectTimeMetric.Set(time.Since(start).Seconds())
+		select {
+		case <-t.C:
+			// re-populate the cache
+			continue
+		case <-ms.ctx.Done():
+			return
+		}
+	}
+}
+
+func (ms *MetricsServer) serveMetricsFromCache() {
+	http.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
+		resp := ms.cachedResponseBody.Load()
+		if resp == nil {
+			return
```

Comment:
I would expect this to return something more meaningful, at least maybe a header asking to come back later?